Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift

```diff
@@ -37,88 +37,444 @@ extension PathHierarchy {
+        // If the next component is an operator, parse the full operator before splitting on "/" ("/" may appear in the operator name)
+        if remaining.unicodeScalars.prefix(3).allSatisfy(\.isValidOperatorHead) {
+            guard let operatorEndIndex = remaining.firstIndex(of: Self.operatorEnd) else {
+                defer { remaining.removeAll() }
+                return remaining
+            }
+            
+            var component = remaining[..<operatorEndIndex]
+            remaining = remaining[operatorEndIndex...]
+            
+            guard let index = remaining.firstIndex(of: Self.separator) else {
+                defer { remaining.removeAll() }
+                return component + remaining
+            }
+            
+            component += remaining[..<index]
+            remaining = remaining[index...].dropFirst() // drop the slash
+            return component
+            
+        }
+        
+        if remaining.first == Self.separator {
+            guard let index = remaining.firstIndex(where: { $0 != Self.separator }) else {
+                defer { remaining.removeAll() }
+                return remaining
+            }
+            let slashPrefix = remaining[..<index]
+            remaining = remaining[index...]
+            return slashPrefix + scanPathComponent()
+        }
+        
+        // If the string doesn't contain a slash then the rest of the string is the component
+        guard let index = remaining.firstIndex(of: Self.separator) else {
+            defer { remaining.removeAll() }
+            return remaining
+        }
+        
+        defer {
+            remaining = remaining[index...].dropFirst() // drop the slash
+        }
+        return remaining[..<index]
+    }
+    
+    mutating func scanAnchorComponent() -> Substring? {
+        guard let index = remaining.firstIndex(of: Self.anchorSeparator) else {
+            return nil
+        }
+        
+        defer { remaining = remaining[..<index] }
+        return remaining[index...].dropFirst() // drop the anchor separator
+    }
+}
+
+private extension StringProtocol {
+    /// Checks if a this string is a valid module name.
+    ///
+    /// Both Swift and Clang require module names to be a valid C99 Extended Identifier.
+    func isValidModuleName() -> Bool {
+        unicodeScalars.allSatisfy {
+            // "-" isn't allowed in module names themselves but it's allowed in link components to separate the disambiguation.
+            $0 == "-" || $0.isValidC99ExtendedIdentifier
+        }
+    }
+}
+
+private extension Unicode.Scalar {
+    /// Checks if this unicode scalar is a valid C99 Extended Identifier.
+    var isValidC99ExtendedIdentifier: Bool {
+        // These is based on "swift-tools-support-core/Sources/TSCUtility/StringMangling.swift"
+        // but that repo is deprecated and not recommended as a dependency.
+        switch value {
+        case
+            // A-Z
+            0x0041...0x005A,
```

Comment:
Any way we can save these unicode values/characters into a config file elsewhere? That would simplify this source code file a lot.

Or would the overhead of opening that config file be too much?

Reply:
In my opinion a separate config file would be more complicated and likely less readable. 

The biggest complication would be that we'd have a resource that we need to ship alongside the `docc` executable. This is a fairly minor thing but it does mean that we need to update `build-script-helper.py` to copy this config file into the install directory and DocC needs to handle the cases when this config file is missing or not formatted correctly. 

There's a tradeoff between readability and ease-of-parsing in a separate file. The easiest to parse would probably be to just write a stream of all the padded bytes for each character but that wouldn't be human readable at all. We could print them as characters which makes it easier to see what characters are supported but it also makes it look like entire Unicode code blocks are supported because most characters in that block are supported. For example, 98 characters in the "Greek" Unicode block are supported C99 identifiers (such as "Ϛ") but 12 characters aren't supported (such as "ϛ"). 

I feel that to make something that's more readable than in-source character ranges with comments we need to write a small one-off parser and tests for that parser

Reply:
Yes you're right a runtime config file would be a bad idea for all the reasons you have here. Maybe just moving these large switch statements into a separate Swift file called something like PathHierarchy+ValidC99ExtendedIdentifier or whatever would remove all the verbose code from here.

Reply:
The downside of moving it to a separate file is that it would need to be `internal` making it available everywhere in DocC whereas now it's only available in this file. That's a very small thing but I like to keep very specialized single use API private in the same file and only move it out if there's other places that need that functionality.

Reply:
Maybe one day I'll find the time to break up `SwiftDocC` into multiple smaller modules and then each piece would have better control over what API it exposes to other pieces 🤞